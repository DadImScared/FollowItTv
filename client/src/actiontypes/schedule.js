
export const ADD_SCHEDULE = 'schedule/ADD_SCHEDULE';
