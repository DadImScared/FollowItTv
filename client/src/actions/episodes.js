
import { ADD_EPISODES } from '../actiontypes/episodes';

export const addEpisodes = (episodes) => {
  return {
    type: ADD_EPISODES,
    episodes
  };
};
