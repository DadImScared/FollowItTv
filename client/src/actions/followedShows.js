
import _ from 'lodash';
import axios from 'axios';

import axiosOptions from './axiosOptions';
import * as followedActionTypes from '../actiontypes/followedShows';
import { addShows } from './shows';

export const addDay = (day, shows) => {
  return {
    type: followedActionTypes.ADD_DAY,
    day,
    shows
  };
};

export const followShow = (days, id) => {
  return {
    type: followedActionTypes.FOLLOW_SHOW,
    days,
    id
  };
};

export const unfollowShow = (days, id) => {
  return {
    type: followedActionTypes.UNFOLLOW_SHOW,
    days,
    id
  };
};

export const postFollow = async (showId) => {
  return await axios.post(`/api/v1/followshow/${showId}/`, {}, axiosOptions());
};

export const addShowIds = (showIdList) => {
  return {
    type: followedActionTypes.ADD_SHOW_IDS,
    showIdList
  };
};

export const extractShowData = (data) => {
  let isArrayOfNumbers;
  const showIds = new Set();
  const showData = {};
  if (typeof data[0] === 'number') {
    isArrayOfNumbers = true;
  }
  else {
    data.forEach((show) => {
      showIds.add(show.id);
      showData[show.id] = show;
    });
  }
  return {
    showData,
    showIds: isArrayOfNumbers ? data:[...showIds]
  };
};

export const getFollowedShows = async (dispatch, day = '') => {
  try {
    const { data } = await axios.get(`/api/v1/followshows/${day ? `${_.capitalize(day)}/`:''}`, axiosOptions());
    const { showData, showIds } = extractShowData(data);
    if (day) {
      dispatch(addShows(showData));
    }
    dispatch(addShowIds(showIds));
  }
  catch (e) {
    throw e;
  }
};
