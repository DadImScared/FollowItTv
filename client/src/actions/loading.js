
import { LOADING_START, LOADING_END } from '../actiontypes/loading';

export const loadingStart = (requestId) => {
  return {
    type: LOADING_START,
    requestId
  };
};

export const loadingEnd = (requestId) => {
  return {
    type: LOADING_END,
    requestId
  };
};
