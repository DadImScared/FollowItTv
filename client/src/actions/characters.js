
import { ADD_CHARACTERS } from '../actiontypes/characters';
import { loadingEnd, loadingStart } from './loading';
import { castSchema } from '../schemas/person';
import { normalize } from 'normalizr';
import { baseUrl } from './seasons';
import axios from 'axios';

export const addCharacters = (entities) => ({
  type: ADD_CHARACTERS,
  ...entities
});

const getCast = async (showId) => {
  return await axios.get(`${baseUrl}shows/${showId}/cast`);
};

export const handleCastData = (data) => {
  const { entities: { person, character }, result } = normalize(data, castSchema);
  const characterIds = result.reduce((acc, { character }) => {
    const newAcc = [...acc];
    if (!newAcc.includes(character)) {
      newAcc.push(character);
    }
    return newAcc;
  }, []);
  return {
    person,
    character,
    result,
    characterIds
  };
};

export const requestCast = (showId) => {
  return async (dispatch) => {
    const loadingId = `GET_CAST_${showId}`;
    dispatch(loadingStart(loadingId));
    try {
      const { data } = await getCast(showId);
      const { person: people, character: characters, characterIds } = handleCastData(data);
      dispatch(addCharacters({ id: showId, people, characters, characterIds }));
      dispatch(loadingEnd(loadingId));
    }
    catch (e) {
      dispatch(loadingEnd(loadingId));
      throw e;
    }
  };
};
