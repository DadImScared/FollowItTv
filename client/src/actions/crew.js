
import axios from 'axios';
import _ from 'lodash';
import { normalize } from 'normalizr';

import { ADD_CREW } from '../actiontypes/crew';
import { loadingStart, loadingEnd } from './loading';
import { baseUrl } from './seasons';
import { crewSchema } from '../schemas/person';

export const addCrew = (entities) => ({
  type: ADD_CREW,
  ...entities
});

const getCrew = async (showId) => await axios.get(`${baseUrl}shows/${showId}/crew`);

export const handleCrewData = (data) => {
  const { entities, result } = normalize(data, crewSchema);
  const crewMembers = result.reduce((acc, crewMember) => {
    return {
      ...acc,
      [crewMember.person]: _.union(_.get(acc, crewMember.person, []), [ crewMember.type ])
    };
  }, {});
  return {
    ...entities,
    result,
    crewMembers
  };
};

export const requestCrew = (showId) => {
  return async (dispatch) => {
    const requestId = `GET_CREW_${showId}`;
    dispatch(loadingStart(requestId));
    try {
      const { data } = await getCrew(showId);
      const { person: people, crewMembers } = handleCrewData(data);
      dispatch(addCrew({ id: showId, people, crewMembers }));
      dispatch(loadingEnd(requestId));
    }
    catch(e) {
      dispatch(loadingEnd(requestId));
      throw e;
    }
  };
};
