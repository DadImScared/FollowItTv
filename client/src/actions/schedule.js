
import axios from 'axios';

import * as scheduleActionTypes from '../actiontypes/schedule';
import { addEpisodes } from './episodes';
import { addShows } from './shows';
import { loadingStart, loadingEnd } from './loading';


export const addSchedule = (date, episodes) => {
  return {
    type: scheduleActionTypes.ADD_SCHEDULE,
    date,
    episodes
  };
};

export const getSchedule = async (date) => await axios.get(`/api/v1/schedule/${date}/`);

export const getEpisodesAndShows = (initialEpisodes) => {
  // we want to remove show info from episodes and replace with show id while also extracting show info
  const episodeIds = [];
  const episodes = {};
  const shows = {};
  initialEpisodes.forEach((obj) => {
    episodeIds.push(obj.id);
    episodes[obj.id] = {
      ...obj,
      show: obj.show.id
    };
    shows[obj.show.id] = {
      ...obj.show
    };
  });
  return {
    episodeIds,
    episodes,
    shows
  };
};

export const requestSchedule = (date) => {
  return async (dispatch, getState) => {
    const { schedule, loading } = getState();
    const requestId = `GET_SCHEDULE_${date}`;
    if (schedule[date] || loading[requestId]) {
      return;
    }
    dispatch(loadingStart(requestId));
    try {
      const { data } = await getSchedule(date);
      const { episodes, shows, episodeIds } = getEpisodesAndShows(data);
      dispatch(addShows(shows));
      dispatch(addEpisodes(episodes));
      dispatch(addSchedule(date, episodeIds));
      dispatch(loadingEnd(requestId));
      return Promise.resolve();
    }
    catch (e) {
      dispatch(loadingEnd(requestId));
      throw e;
    }
  };
};
