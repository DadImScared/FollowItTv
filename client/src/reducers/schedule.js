
import * as scheduleActionTypes from '../actiontypes/schedule';


function schedule(state={}, action) {
  switch(action.type) {
  case scheduleActionTypes.ADD_SCHEDULE:
    return {
      ...state,
      [action.date]: action.episodes
    };
  default:
    return state;
  }
}

export default schedule;
