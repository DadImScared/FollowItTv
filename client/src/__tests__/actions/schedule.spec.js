
import axios from 'axios';

import * as scheduleActionTypes from '../../actiontypes/schedule';
import * as actions from '../../actions/schedule';

jest.mock('axios');

// actions.getSchedule = jest.fn((date) => response[date]);

const shows = [
  {
    id: 1,
    name: 'breaking bad'
  },
  {
    id: 3,
    name: 'under the dome'
  }
];

const detailedResponse = {
  data: [
    {
      id: 1,
      name: 'episode0',
      show: shows[0]
    },
    {
      id: 2,
      name: 'episode2',
      show: shows[1]
    }
  ]
};

describe('schedule actions', () => {
  it('should create add schedule action', () => {
    const date = '2018-03-17';
    const episodes = [{ name: 'episode1' }];
    const expectedAction = {
      type: scheduleActionTypes.ADD_SCHEDULE,
      date,
      episodes
    };
    expect(actions.addSchedule(date, episodes)).toEqual(expectedAction);
  });

  it('should get schedule', async () => {
    const response = {
      data: [{ name: 'episode0' }]
    };
    axios.get.mockImplementation(() => response);
    const { data } = await actions.getSchedule('2018-03-17');
    expect(data).toEqual(response.data);
  });

  it('should get episodes and shows from initial episodes', async () => {
    axios.get.mockImplementation(() => detailedResponse);
    const { data } = await actions.getSchedule('2018-03-17');
    const { episodes, shows: newShows, episodeIds } = actions.getEpisodesAndShows(data);
    expect(episodeIds).toEqual([1, 2]);
    expect(episodes).toEqual({
      1: {
        id: 1,
        name: 'episode0',
        show: 1
      },
      2: {
        id: 2,
        name: 'episode2',
        show: 3
      }
    });
    expect(newShows).toEqual({
      [shows[0].id]: shows[0],
      [shows[1].id]: shows[1]
    });
  });
});
