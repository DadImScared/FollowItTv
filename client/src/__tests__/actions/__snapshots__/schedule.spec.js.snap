// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`schedule actions get schedule 1`] = `
Array [
  Object {
    "requestId": "GET_SCHEDULE_2011-03-26",
    "type": "loading/LOADING_START",
  },
  Object {
    "shows": Object {
      "1": Object {
        "id": 1,
        "name": "breaking bad",
      },
      "3": Object {
        "id": 3,
        "name": "under the dome",
      },
    },
    "type": "shows/ADD_SHOWS",
  },
  Object {
    "episodes": Object {
      "1": Object {
        "id": 1,
        "name": "episode0",
        "show": 1,
      },
      "2": Object {
        "id": 2,
        "name": "episode2",
        "show": 3,
      },
    },
    "type": "episodes/ADD_EPISODES",
  },
  Object {
    "date": "2011-03-26",
    "episodes": Array [
      1,
      2,
    ],
    "type": "schedule/ADD_SCHEDULE",
  },
  Object {
    "requestId": "GET_SCHEDULE_2011-03-26",
    "type": "loading/LOADING_END",
  },
]
`;

exports[`schedule actions get schedule error 1`] = `
Array [
  Object {
    "requestId": "GET_SCHEDULE_2011-03-26",
    "type": "loading/LOADING_START",
  },
  Object {
    "requestId": "GET_SCHEDULE_2011-03-26",
    "type": "loading/LOADING_END",
  },
]
`;
